// @ciscospark/plugin-mercury/dist/errors.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.AuthorizationError = exports.ConnectionError = undefined;

var _defineProperties = require('babel-runtime/core-js/object/define-properties');

var _defineProperties2 = _interopRequireDefault(_defineProperties);

var _extendError = require('extend-error');

var _extendError2 = _interopRequireDefault(_extendError);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var ConnectionError = exports.ConnectionError = (0, _extendError2.default)({
  parseFn: function parseFn(event) {
    (0, _defineProperties2.default)(this, {
      code: {
        value: event.code
      },
      reason: {
        value: event.reason
      }
    });

    return event.reason;
  },


  subTypeName: 'ConnectionError'
}); /**!
     *
     * Copyright (c) 2015-2016 Cisco Systems, Inc. See LICENSE file.
     */

var AuthorizationError = exports.AuthorizationError = (0, _extendError2.default)(ConnectionError, {
  subTypeName: 'AuthorizationError'
});
