// @ciscospark/plugin-metrics/dist/metrics.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _getOwnPropertyDescriptor = require('babel-runtime/core-js/object/get-own-property-descriptor');

var _getOwnPropertyDescriptor2 = _interopRequireDefault(_getOwnPropertyDescriptor);

var _assign = require('babel-runtime/core-js/object/assign');

var _assign2 = _interopRequireDefault(_assign);

var _dec, _desc, _value, _obj; /**!
                                *
                                * Copyright (c) 2015-2016 Cisco Systems, Inc. See LICENSE file.
                                * @private
                                */

var _sparkCore = require('@ciscospark/spark-core');

var _batcher = require('./batcher');

var _batcher2 = _interopRequireDefault(_batcher);

var _clientMetricsBatcher = require('./client-metrics-batcher');

var _clientMetricsBatcher2 = _interopRequireDefault(_clientMetricsBatcher);

var _coreDecorators = require('core-decorators');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _applyDecoratedDescriptor(target, property, decorators, descriptor, context) {
  var desc = {};
  Object['ke' + 'ys'](descriptor).forEach(function (key) {
    desc[key] = descriptor[key];
  });
  desc.enumerable = !!desc.enumerable;
  desc.configurable = !!desc.configurable;

  if ('value' in desc || desc.initializer) {
    desc.writable = true;
  }

  desc = decorators.slice().reverse().reduce(function (desc, decorator) {
    return decorator(target, property, desc) || desc;
  }, desc);

  if (context && desc.initializer !== void 0) {
    desc.value = desc.initializer ? desc.initializer.call(context) : void 0;
    desc.initializer = undefined;
  }

  if (desc.initializer === void 0) {
    Object['define' + 'Property'](target, property, desc);
    desc = null;
  }

  return desc;
}

var Metrics = _sparkCore.SparkPlugin.extend((_dec = (0, _coreDecorators.deprecated)('Metrics#sendUnstructured() is deprecated; please use Metrics#submit()'), (_obj = {
  children: {
    batcher: _batcher2.default,
    clientMetricsBatcher: _clientMetricsBatcher2.default
  },

  namespace: 'Metrics',

  sendUnstructured: function sendUnstructured(key, value) {
    return this.submit(key, value);
  },
  submit: function submit(key, value) {
    return this.batcher.request((0, _assign2.default)({ key: key }, value));
  },
  submitClientMetrics: function submitClientMetrics(eventName, props) {
    var payload = { metricName: eventName };
    if (props.tags) {
      payload.tags = props.tags;
    }
    if (props.fields) {
      payload.fields = props.fields;
    }
    return this.clientMetricsBatcher.request(payload);
  }
}, (_applyDecoratedDescriptor(_obj, 'sendUnstructured', [_dec], (0, _getOwnPropertyDescriptor2.default)(_obj, 'sendUnstructured'), _obj)), _obj)));

exports.default = Metrics;
