// @ciscospark/plugin-wdm/dist/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.UrlInterceptor = exports.EmbargoInterceptor = exports.AuthInterceptor = exports.FeatureModel = exports.FeatureCollection = exports.FeaturesModel = exports.Device = exports.default = undefined;

var _device = require('./device');

Object.defineProperty(exports, 'default', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_device).default;
  }
});
Object.defineProperty(exports, 'Device', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_device).default;
  }
});

var _featuresModel = require('./device/features-model');

Object.defineProperty(exports, 'FeaturesModel', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_featuresModel).default;
  }
});

var _featureCollection = require('./device/feature-collection');

Object.defineProperty(exports, 'FeatureCollection', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_featureCollection).default;
  }
});

var _featureModel = require('./device/feature-model');

Object.defineProperty(exports, 'FeatureModel', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_featureModel).default;
  }
});

var _auth = require('./interceptors/auth');

Object.defineProperty(exports, 'AuthInterceptor', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_auth).default;
  }
});

var _embargo = require('./interceptors/embargo');

Object.defineProperty(exports, 'EmbargoInterceptor', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_embargo).default;
  }
});

var _url = require('./interceptors/url');

Object.defineProperty(exports, 'UrlInterceptor', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_url).default;
  }
});

var _sparkCore = require('@ciscospark/spark-core');

var _device2 = _interopRequireDefault(_device);

var _config = require('./config');

var _config2 = _interopRequireDefault(_config);

var _url2 = _interopRequireDefault(_url);

var _auth2 = _interopRequireDefault(_auth);

var _embargo2 = _interopRequireDefault(_embargo);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**!
 *
 * Copyright (c) 2015-2016 Cisco Systems, Inc. See LICENSE file.
 */

(0, _sparkCore.registerPlugin)('device', _device2.default, {
  interceptors: {
    UrlInterceptor: _url2.default.create,
    AuthInterceptor: _auth2.default.create,
    EmbargoInterceptor: _embargo2.default.create
  },
  config: _config2.default
});
