// @ciscospark/spark-core/dist/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Page = exports.Batcher = exports.grantErrors = exports.Credentials = exports.AuthInterceptor = exports.Authorization = exports.SparkTrackingIdInterceptor = exports.RequestTimingInterceptor = exports.RequestLoggerInterceptor = exports.ResponseLoggerInterceptor = exports.RedirectInterceptor = exports.PayloadTransformerInterceptor = exports.NetworkTimingInterceptor = exports.SparkPlugin = exports.SparkHttpError = exports.registerPlugin = exports.default = exports.children = exports.waitForValue = exports.StorageError = exports.Store = exports.SparkPluginStorage = exports.persist = exports.NotFoundError = exports.MemoryStoreAdapter = exports.makeSparkPluginStore = exports.makeSparkStore = undefined;

var _storage = require('./lib/storage');

Object.defineProperty(exports, 'makeSparkStore', {
  enumerable: true,
  get: function get() {
    return _storage.makeSparkStore;
  }
});
Object.defineProperty(exports, 'makeSparkPluginStore', {
  enumerable: true,
  get: function get() {
    return _storage.makeSparkPluginStore;
  }
});
Object.defineProperty(exports, 'MemoryStoreAdapter', {
  enumerable: true,
  get: function get() {
    return _storage.MemoryStoreAdapter;
  }
});
Object.defineProperty(exports, 'NotFoundError', {
  enumerable: true,
  get: function get() {
    return _storage.NotFoundError;
  }
});
Object.defineProperty(exports, 'persist', {
  enumerable: true,
  get: function get() {
    return _storage.persist;
  }
});
Object.defineProperty(exports, 'SparkPluginStorage', {
  enumerable: true,
  get: function get() {
    return _storage.SparkPluginStorage;
  }
});
Object.defineProperty(exports, 'Store', {
  enumerable: true,
  get: function get() {
    return _storage.Store;
  }
});
Object.defineProperty(exports, 'StorageError', {
  enumerable: true,
  get: function get() {
    return _storage.StorageError;
  }
});
Object.defineProperty(exports, 'waitForValue', {
  enumerable: true,
  get: function get() {
    return _storage.waitForValue;
  }
});

var _sparkCore = require('./spark-core');

Object.defineProperty(exports, 'children', {
  enumerable: true,
  get: function get() {
    return _sparkCore.children;
  }
});
Object.defineProperty(exports, 'default', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_sparkCore).default;
  }
});
Object.defineProperty(exports, 'registerPlugin', {
  enumerable: true,
  get: function get() {
    return _sparkCore.registerPlugin;
  }
});

var _sparkHttpError = require('./lib/spark-http-error');

Object.defineProperty(exports, 'SparkHttpError', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_sparkHttpError).default;
  }
});

var _sparkPlugin = require('./lib/spark-plugin');

Object.defineProperty(exports, 'SparkPlugin', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_sparkPlugin).default;
  }
});

var _networkTiming = require('./interceptors/network-timing');

Object.defineProperty(exports, 'NetworkTimingInterceptor', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_networkTiming).default;
  }
});

var _payloadTransformer = require('./interceptors/payload-transformer');

Object.defineProperty(exports, 'PayloadTransformerInterceptor', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_payloadTransformer).default;
  }
});

var _redirect = require('./interceptors/redirect');

Object.defineProperty(exports, 'RedirectInterceptor', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_redirect).default;
  }
});

var _responseLogger = require('./interceptors/response-logger');

Object.defineProperty(exports, 'ResponseLoggerInterceptor', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_responseLogger).default;
  }
});

var _requestLogger = require('./interceptors/request-logger');

Object.defineProperty(exports, 'RequestLoggerInterceptor', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_requestLogger).default;
  }
});

var _requestTiming = require('./interceptors/request-timing');

Object.defineProperty(exports, 'RequestTimingInterceptor', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_requestTiming).default;
  }
});

var _sparkTrackingId = require('./interceptors/spark-tracking-id');

Object.defineProperty(exports, 'SparkTrackingIdInterceptor', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_sparkTrackingId).default;
  }
});

var _credentials = require('./plugins/credentials');

Object.defineProperty(exports, 'Authorization', {
  enumerable: true,
  get: function get() {
    return _credentials.Authorization;
  }
});
Object.defineProperty(exports, 'AuthInterceptor', {
  enumerable: true,
  get: function get() {
    return _credentials.AuthInterceptor;
  }
});
Object.defineProperty(exports, 'Credentials', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_credentials).default;
  }
});
Object.defineProperty(exports, 'grantErrors', {
  enumerable: true,
  get: function get() {
    return _credentials.grantErrors;
  }
});

var _batcher = require('./lib/batcher');

Object.defineProperty(exports, 'Batcher', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_batcher).default;
  }
});

var _page = require('./lib/page');

Object.defineProperty(exports, 'Page', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_page).default;
  }
});

require('./plugins/logger');

require('./plugins/credentials');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
