// @ciscospark/plugin-wdm/dist/device/device.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _getOwnPropertyDescriptor = require('babel-runtime/core-js/object/get-own-property-descriptor');

var _getOwnPropertyDescriptor2 = _interopRequireDefault(_getOwnPropertyDescriptor);

var _omit = require('lodash/omit');

var _omit2 = _interopRequireDefault(_omit);

var _apply = require('babel-runtime/core-js/reflect/apply');

var _apply2 = _interopRequireDefault(_apply);

var _keys = require('babel-runtime/core-js/object/keys');

var _keys2 = _interopRequireDefault(_keys);

var _getIterator2 = require('babel-runtime/core-js/get-iterator');

var _getIterator3 = _interopRequireDefault(_getIterator2);

var _promise = require('babel-runtime/core-js/promise');

var _promise2 = _interopRequireDefault(_promise);

var _dec, _dec2, _dec3, _dec4, _dec5, _dec6, _dec7, _dec8, _desc, _value, _obj; /**!
                                                                                 *
                                                                                 * Copyright (c) 2015-2016 Cisco Systems, Inc. See LICENSE file.
                                                                                 */

var _common = require('@ciscospark/common');

var _sparkCore = require('@ciscospark/spark-core');

var _util = require('util');

var _util2 = _interopRequireDefault(_util);

var _featuresModel = require('./features-model');

var _featuresModel2 = _interopRequireDefault(_featuresModel);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _applyDecoratedDescriptor(target, property, decorators, descriptor, context) {
  var desc = {};
  Object['ke' + 'ys'](descriptor).forEach(function (key) {
    desc[key] = descriptor[key];
  });
  desc.enumerable = !!desc.enumerable;
  desc.configurable = !!desc.configurable;

  if ('value' in desc || desc.initializer) {
    desc.writable = true;
  }

  desc = decorators.slice().reverse().reduce(function (desc, decorator) {
    return decorator(target, property, desc) || desc;
  }, desc);

  if (context && desc.initializer !== void 0) {
    desc.value = desc.initializer ? desc.initializer.call(context) : void 0;
    desc.initializer = undefined;
  }

  if (desc.initializer === void 0) {
    Object['define' + 'Property'](target, property, desc);
    desc = null;
  }

  return desc;
}

var Device = _sparkCore.SparkPlugin.extend((_dec = (0, _sparkCore.waitForValue)('@'), _dec2 = (0, _sparkCore.waitForValue)('@'), _dec3 = (0, _sparkCore.waitForValue)('@'), _dec4 = (0, _sparkCore.waitForValue)('@'), _dec5 = (0, _sparkCore.waitForValue)('@'), _dec6 = (0, _sparkCore.waitForValue)('@'), _dec7 = (0, _sparkCore.waitForValue)('@'), _dec8 = (0, _sparkCore.waitForValue)('@'), (_obj = {
  children: {
    features: _featuresModel2.default
  },

  idAttribute: 'url',

  namespace: 'Device',

  props: {
    intranetInactivityDuration: 'number',
    intranetInactivityCheckUrl: 'string',
    modificationTime: 'string',
    searchEncryptionKeyUrl: 'string',
    services: {
      // Even though @jodykstr will tell you the docs claim you don't need to
      // initialize `object` properties, the docs lie.
      default: function _default() {
        return {};
      },

      type: 'object'
    },
    url: 'string',
    userId: 'string',
    webSocketUrl: 'string'
  },

  derived: {
    registered: {
      deps: ['url'],
      fn: function fn() {
        return Boolean(this.url);
      }
    }
  },

  session: {
    // Fun Fact: setTimeout returns a Timer object instead of a Number in Node 6
    logoutTimer: 'any',
    lastUserActivityDate: 'number'
  },

  determineService: function determineService(url) {
    var _iteratorNormalCompletion = true;
    var _didIteratorError = false;
    var _iteratorError = undefined;

    try {
      for (var _iterator = (0, _getIterator3.default)((0, _keys2.default)(this.services)), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
        var key = _step.value;

        var serviceUrl = this.services[key];
        if (url.startsWith(serviceUrl)) {
          // "ServiceUrl" is 10 characters
          return _promise2.default.resolve(key.substr(0, key.length - 10));
        }
      }
    } catch (err) {
      _didIteratorError = true;
      _iteratorError = err;
    } finally {
      try {
        if (!_iteratorNormalCompletion && _iterator.return) {
          _iterator.return();
        }
      } finally {
        if (_didIteratorError) {
          throw _iteratorError;
        }
      }
    }

    return _promise2.default.reject(new Error(url + ' does not reflect a known service'));
  },
  getServiceUrl: function getServiceUrl(service) {
    var _this = this;

    return this._getServiceUrl(this.services, service).then(function (isServiceUrl) {
      return isServiceUrl || _this.getPreDiscoveryServiceUrl(service);
    });
  },
  getPreDiscoveryServiceUrl: function getPreDiscoveryServiceUrl(service) {
    // The Promise.resolve here is temporary. A future PR will make the
    // corresponding _ method async to allow for lazy device registration
    return _promise2.default.resolve(this._getServiceUrl(this.config.preDiscoveryServices, service));
  },
  initialize: function initialize() {
    var _this2 = this;

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    (0, _apply2.default)(_sparkCore.SparkPlugin.prototype.initialize, this, args);

    // Propagate change(:[attribute]) events from collections
    ['developer', 'entitlement', 'user'].forEach(function (collectionName) {
      _this2.features.on('change:' + collectionName, function (model, value, options) {
        _this2.trigger('change', _this2, options);
        _this2.trigger('change:features', _this2, _this2.features, options);
      });
    });

    this.listenToAndRun(this, 'change:intranetInactivityCheckUrl', function () {
      return _this2._resetLogoutTimer();
    });
    this.listenToAndRun(this, 'change:intranetInactivityDuration', function () {
      return _this2._resetLogoutTimer();
    });
    this.listenTo(this.spark, 'user-activity', function () {
      _this2.lastUserActivityDate = Date.now();
    });
  },


  /**
   * Don't log the features object
   * @param {number} depth
   * @returns {Object}
   */
  inspect: function inspect(depth) {
    return _util2.default.inspect((0, _omit2.default)(this.serialize(), 'features'), { depth: depth });
  },
  isPreDiscoveryService: function isPreDiscoveryService(service) {
    // The Promise.resolve here is temporary. A future PR will make the
    // corresponding _ method async to allow for lazy device registration
    return _promise2.default.resolve(this._isService(this.config.preDiscoveryServices, service));
  },
  isPreDiscoveryServiceUrl: function isPreDiscoveryServiceUrl(uri) {
    // The Promise.resolve here is temporary. A future PR will make the
    // corresponding _ method async to allow for lazy device registration
    return _promise2.default.resolve(this._isServiceUrl(this.config.preDiscoveryServices, uri));
  },
  isService: function isService(service) {
    var _this3 = this;

    return this._isService(this.services, service).then(function (_isService) {
      return _isService || _this3.isPreDiscoveryService(service);
    });
  },
  isServiceUrl: function isServiceUrl(uri) {
    // The Promise.resolve here is temporary. A future PR will make the
    // corresponding _ method async to allow for lazy device registration
    return _promise2.default.resolve(this._isServiceUrl(this.services, uri));
  },
  isSpecificService: function isSpecificService(service, key) {
    if (key === service) {
      return _promise2.default.resolve(true);
    }

    return this.getServiceUrl(service).then(function (serviceUrl) {
      return key.includes(serviceUrl);
    });
  },
  _getServiceUrl: function _getServiceUrl(target, service) {
    /* istanbul ignore if */
    if (!target) {
      return _promise2.default.reject(new Error('`target` is a required parameter'));
    }

    if (!service) {
      return _promise2.default.reject(new Error('`service` is a required parameter'));
    }

    return _promise2.default.resolve(target[service + 'ServiceUrl']);
  },
  _isService: function _isService(target, service) {
    return this._getServiceUrl(target, service).then(function (url) {
      return Boolean(url);
    });
  },
  _isServiceUrl: function _isServiceUrl(target, uri) {
    if (!uri) {
      return _promise2.default.reject(new Error('`uri` is a required parameter'));
    }

    for (var key in target) {
      if (target[key] && uri.indexOf(target[key]) === 0) {
        return _promise2.default.resolve(true);
      }
    }

    return _promise2.default.resolve(false);
  },
  refresh: function refresh() {
    var _this4 = this;

    this.logger.info('device: refreshing');

    if (!this.registered) {
      this.logger.info('device: device not registered, refreshing');
      return this.register();
    }

    this.logger.info('device: refreshing');

    return this.request({
      method: 'PUT',
      uri: this.url,
      body: (0, _omit2.default)(this.serialize(), 'features', 'mediaClusters')
    }).then(function (res) {
      return _this4._processRegistrationSuccess(res);
    }).catch(function (reason) {
      if (reason.statusCode === 404) {
        // If we get a 404, it means the device is no longer valid and we need
        // to register a new one.
        _this4.logger.info('device: refresh failed with 404, attempting to register new device');
        _this4.clear();
        return _this4.register();
      }
      return _promise2.default.reject(reason);
    });
  },
  register: function register() {
    var _this5 = this;

    /* eslint no-invalid-this: [0] */
    this.logger.info('device: registering');

    if (this.registered) {
      this.logger.info('device: device already registered, refreshing');
      return this.refresh();
    }

    return this.request({
      method: 'POST',
      service: 'wdm',
      resource: 'devices',
      body: this.config.defaults
    }).then(function (res) {
      return _this5._processRegistrationSuccess(res);
    });
  },
  unregister: function unregister() {
    var _this6 = this;

    this.logger.info('device: unregistering');

    if (!this.url) {
      throw new Error('device: not registered');
    }

    return this.request({
      uri: this.url,
      method: 'DELETE'
    }).then(function () {
      return _this6.clear();
    });
  },
  _processRegistrationSuccess: function _processRegistrationSuccess(res) {
    this.logger.info('device: received registration payload');
    this.set(res.body);
  },
  _resetLogoutTimer: function _resetLogoutTimer() {
    var _this7 = this;

    clearTimeout(this.logoutTimer);
    this.unset('logoutTimer');
    if (this.intranetInactivityCheckUrl && this.intranetInactivityDuration) {
      this.on('change:lastUserActivityDate', function () {
        return _this7._resetLogoutTimer();
      });

      var timer = setTimeout(function () {
        _this7.spark.request({
          method: 'GET',
          uri: _this7.intranetInactivityCheckUrl
        }).catch(function () {
          _this7.logger.info('device: did not reach internal ping endpoint; logging out after inactivity on a public network');
          return _this7.spark.logout();
        }).catch(function (reason) {
          _this7.logger.warn('device: logout failed', reason);
        });
      }, this.intranetInactivityDuration * 1000);

      this.logoutTimer = timer;
    }
  }
}, (_applyDecoratedDescriptor(_obj, 'determineService', [_dec], (0, _getOwnPropertyDescriptor2.default)(_obj, 'determineService'), _obj), _applyDecoratedDescriptor(_obj, 'getServiceUrl', [_dec2], (0, _getOwnPropertyDescriptor2.default)(_obj, 'getServiceUrl'), _obj), _applyDecoratedDescriptor(_obj, 'initialize', [_sparkCore.persist], (0, _getOwnPropertyDescriptor2.default)(_obj, 'initialize'), _obj), _applyDecoratedDescriptor(_obj, 'isService', [_dec3], (0, _getOwnPropertyDescriptor2.default)(_obj, 'isService'), _obj), _applyDecoratedDescriptor(_obj, 'isServiceUrl', [_dec4], (0, _getOwnPropertyDescriptor2.default)(_obj, 'isServiceUrl'), _obj), _applyDecoratedDescriptor(_obj, 'isSpecificService', [_dec5], (0, _getOwnPropertyDescriptor2.default)(_obj, 'isSpecificService'), _obj), _applyDecoratedDescriptor(_obj, 'refresh', [_common.oneFlight, _dec6], (0, _getOwnPropertyDescriptor2.default)(_obj, 'refresh'), _obj), _applyDecoratedDescriptor(_obj, 'register', [_common.oneFlight, _dec7], (0, _getOwnPropertyDescriptor2.default)(_obj, 'register'), _obj), _applyDecoratedDescriptor(_obj, 'unregister', [_common.oneFlight, _dec8], (0, _getOwnPropertyDescriptor2.default)(_obj, 'unregister'), _obj)), _obj)));

exports.default = Device;
