// @ciscospark/spark-core/dist/plugins/credentials/auth-interceptor.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _slicedToArray2 = require('babel-runtime/helpers/slicedToArray');

var _slicedToArray3 = _interopRequireDefault(_slicedToArray2);

var _promise = require('babel-runtime/core-js/promise');

var _promise2 = _interopRequireDefault(_promise);

var _deleteProperty = require('babel-runtime/core-js/reflect/delete-property');

var _deleteProperty2 = _interopRequireDefault(_deleteProperty);

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _httpCore = require('@ciscospark/http-core');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * @class
 */
var AuthInterceptor = function (_Interceptor) {
  (0, _inherits3.default)(AuthInterceptor, _Interceptor);

  function AuthInterceptor() {
    (0, _classCallCheck3.default)(this, AuthInterceptor);
    return (0, _possibleConstructorReturn3.default)(this, (AuthInterceptor.__proto__ || (0, _getPrototypeOf2.default)(AuthInterceptor)).apply(this, arguments));
  }

  (0, _createClass3.default)(AuthInterceptor, [{
    key: 'onRequest',


    /**
     * @see Interceptor#onRequest
     * @param {Object} options
     * @returns {Object}
     */
    value: function onRequest(options) {
      var _this2 = this;

      // If Authorizations is already set, don't overwrite it
      if ('authorization' in options.headers) {
        // If Authorization is set to null, false, or undefined, delete it to
        // prevent a CORS preflight.
        if (!options.headers.authorization) {
          (0, _deleteProperty2.default)(options.headers, 'authorization');
        }
        return options;
      }

      return _promise2.default.all([this.requiresCredentials(options), this.requiresClientCredentials(options)]).then(function (results) {
        var _results = (0, _slicedToArray3.default)(results, 2),
            requiresCredentials = _results[0],
            requiresClientCredentials = _results[1];

        if (requiresClientCredentials) {
          return _this2.spark.credentials.getClientCredentialsAuthorization();
        } else if (requiresCredentials) {
          return _this2.spark.credentials.getAuthorization();
        }
        return null;
      }).then(function (authorization) {
        options.headers.authorization = authorization;
        return options;
      });
    }

    /**
     * @see Interceptor#onResponseError
     * @param {Object} options
     * @param {Error} reason
     * @returns {Object}
     */

  }, {
    key: 'onResponseError',
    value: function onResponseError(options, reason) {
      var _this3 = this;

      return this.shouldAttemptReauth(reason, options).then(function (shouldAttemptReauth) {
        if (shouldAttemptReauth) {
          _this3.spark.logger.info('auth: received 401, attempting to reauthenticate');

          if (reason.options.headers) {
            (0, _deleteProperty2.default)(reason.options.headers, 'authorization');
          }

          return _this3.spark.refresh({ force: true }).then(function () {
            return _this3.replay(options);
          });
        }

        return _promise2.default.reject(reason);
      });
    }

    /**
     * Replays the request
     * @param {Object} options
     * @returns {Object}
     */

  }, {
    key: 'replay',
    value: function replay(options) {
      options.replayCount = options.replayCount || 1;

      this.spark.logger.info('auth: replaying request ' + options.replayCount + ' time');

      if (options.replayCount > this.spark.config.maxReplayAttempts) {
        this.spark.logger.error('auth: failed after ' + this.spark.config.maxReplayAttempts + ' replay attempts');
        return _promise2.default.reject(new Error('Failed after ' + this.spark.config.maxReplayAttempts + ' replay attempts'));
      }

      return this.spark.request(options);
    }

    /**
     * Indicates whether or not the current request requires credentials
     * @param {Object} options
     * @returns {Promise<boolean>}
     */

  }, {
    key: 'requiresCredentials',
    value: function requiresCredentials(options) {
      this.spark.logger.warn('AuthInterceptor: #requiresCredentials should be overridden');
      if (options.uri.includes(this.spark.config.credentials.oauth.tokenUrl)) {
        return _promise2.default.resolve(false);
      }

      if (options.uri.includes(this.spark.config.credentials.oauth.authorizeUrl)) {
        return _promise2.default.resolve(false);
      }

      if (options.uri.includes(this.spark.config.credentials.samlUrl)) {
        return _promise2.default.resolve(false);
      }

      return _promise2.default.resolve(true);
    }

    /**
     * Indicates whether or not the current request requires client credentials.
     * @param {Object} options
     * @returns {Promise<boolean>}
     */

  }, {
    key: 'requiresClientCredentials',
    value: function requiresClientCredentials(options) {
      return options.requiresClientCredentials;
    }

    /**
     * Indicates whether or not the current request should refresh its access
     * token in event of a 401
     * @param {Error} reason
     * @param {Object} options
     * @returns {Promise<boolean>}
     */

  }, {
    key: 'shouldAttemptReauth',
    value: function shouldAttemptReauth(reason, options) {
      if (options && options.shouldRefreshAccessToken === false) {
        return _promise2.default.resolve(false);
      }

      if (options && options.requiresClientCredentials) {
        return _promise2.default.resolve(false);
      }

      if (reason.statusCode === 401) {
        return _promise2.default.resolve(true);
      }

      return _promise2.default.resolve(false);
    }
  }], [{
    key: 'create',

    /**
     * @returns {AuthInterceptor}
     */
    value: function create() {
      return new AuthInterceptor({ spark: this });
    }
  }]);
  return AuthInterceptor;
}(_httpCore.Interceptor); /**!
                           *
                           * Copyright (c) 2015 Cisco Systems, Inc. See LICENSE file.
                           * @private
                           */

exports.default = AuthInterceptor;
