// @ciscospark/spark-core/dist/plugins/credentials/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.AuthInterceptor = exports.grantErrors = exports.Authorization = exports.Credentials = exports.default = undefined;

var _authorization = require('./authorization');

Object.defineProperty(exports, 'Authorization', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_authorization).default;
  }
});

var _grantErrors = require('./grant-errors');

Object.defineProperty(exports, 'grantErrors', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_grantErrors).default;
  }
});

var _sparkCore = require('../../spark-core');

var _authInterceptor = require('./auth-interceptor');

var _authInterceptor2 = _interopRequireDefault(_authInterceptor);

var _credentials = require('./credentials');

var _credentials2 = _interopRequireDefault(_credentials);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

(0, _sparkCore.registerPlugin)('credentials', _credentials2.default, {
  proxies: ['isAuthenticated', 'isAuthenticating'],
  interceptors: {
    AuthInterceptor: _authInterceptor2.default.create
  }
}); /**!
     *
     * Copyright (c) 2015 Cisco Systems, Inc. See LICENSE file.
     * @private
     */

exports.default = _credentials2.default;
exports.Credentials = _credentials2.default;
exports.AuthInterceptor = _authInterceptor2.default;
